### PR TITLE
More featureful lpc mode in indent.

### PR DESCRIPTION
**Pull request by rjsalts**
lpc mode now understands closures and symbols.
It will also avoid treating a colon as a label for goto,
outdenting only for case statements in a switch.
Using va_args for error printing.

### CODE REVIEW

File: src/util/indent/lexi.c [hunk 1]
```diff
@@ -335,6 +337,10 @@ lexi()
 		  break;
 	      }
 	    buf_ptr++;
+	    if (lpc && qchar == '\'' && !(isalnum(*buf_ptr)||*buf_ptr=='_')) {
```
Comment (amotzkau) [inline src/util/indent/lexi.c]: Could you please add a comment, what this condition checks for?
Also pay attention that buf_ptr might point after the end of the buffer (the condition is between incrementation of the pointer and the check for the end of the buffer), so dereferencing might not be a good idea.

File: src/util/indent/lexi.c [hunk 2]
```diff
@@ -372,6 +378,19 @@ lexi()
 	break;
 
     case '#':
+	if (lpc && *buf_ptr == '\'' ) { /* it's a closure */
+	    buf_ptr++;
+	    while(!(isspace(*buf_ptr)||*buf_ptr=='{'||*buf_ptr=='}'))
```
Comment (amotzkau) [inline src/util/indent/lexi.c]: The dereference of buf_ptr is done without checking that it's still within the buffer.

File: src/util/indent/lexi.c [hunk 3]
```diff
@@ -372,6 +378,19 @@ lexi()
...
+	      {
+	        buf_ptr++;
+		if (*buf_ptr == ',' )
```
Comment (amotzkau) [inline src/util/indent/lexi.c]: Same here, dereferencing buf_ptr without checking the bounds.

File: src/util/indent/lexi.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+enum rwcodes {
```
Comment (amotzkau) [inline src/util/indent/lexi.h]: Would you please remove the enum from the c file, so we don't have a duplicate definition of the enum?

File: src/util/indent/lexi.c (hunk 2 above)
Comment (rjsalts) [reply src/util/indent/lexi.c]: My C is not good. I take it I need to test on buf_end somehow, would it be enough to reorder the body of the while loop to do if(buf_ptr >= buf_end) fill_buffer(); before the if(*buf_ptr == ',') or do I have a problem outside the while loop on line 382?

**Comment by amotzkau** [comment]
I have splitted the commits to be more clearly laid out and merged them.

**Pull request closed without merging**